spec/ref_arch_setup/download_helper_spec.rb

```diff
@@ -0,0 +1,933 @@
+
+          expect(subject).to receive(:valid_response?)
+            .with(test_http_response).and_return(true)
+
+          expect(test_http_response).to receive(:body).and_return(test_body)
+
+          expect(test_oga).to receive(:parse_html)
+            .with(test_body).and_return(test_document)
+
+          expect(test_document).to receive(:xpath)
+            .with(xpath).and_return(test_result)
+
+          expect(subject.parse_prod_versions_url(xpath)).to eq(test_result)
+        end
+      end
+
+      context "when an xpath is not specified" do
+        it "uses the default xpath and returns the versions" do
+          xpath = "//table/tbody/tr/td[1]"
+
+          expect(subject).to receive(:puts).with(message)
+
+          expect(test_uri).to receive(:parse)
+            .with(TEST_PE_VERSIONS_URL).and_return(test_pe_versions_uri)
+
+          expect(test_net_http).to receive(:get_response)
+            .with(test_pe_versions_uri).and_return(test_http_response)
+
+          expect(subject).to receive(:valid_response?)
+            .with(test_http_response).and_return(true)
+
+          expect(test_http_response).to receive(:body).and_return(test_body)
+
+          expect(test_oga).to receive(:parse_html)
+            .with(test_body).and_return(test_document)
+
+          expect(test_document).to receive(:xpath)
+            .with(xpath).and_return(test_result)
+
+          expect(subject.parse_prod_versions_url).to eq(test_result)
+        end
+      end
+    end
+
+    context "when the response is not valid" do
+      it "raises an error" do
+        error = "Invalid response"
+
+        expect(subject).to receive(:puts).with(message)
+
+        expect(test_uri).to receive(:parse)
+          .with(TEST_PE_VERSIONS_URL).and_return(test_pe_versions_uri)
+
+        expect(test_net_http).to receive(:get_response)
+          .with(test_pe_versions_uri).and_return(test_http_response)
+
+        expect(subject).to receive(:valid_response?)
+          .with(test_http_response).and_return(false)
+
+        expect(test_http_response).not_to receive(:body)
+
+        expect { subject.parse_prod_versions_url }.to raise_error(RuntimeError, error)
+      end
+    end
+  end
+
+  describe "#valid_response?" do
+    let(:test_http_response) { Class.new }
+
+    context "when a valid response is provided" do
+      code = TEST_VALID_RESPONSE_CODE
+      body = TEST_VALID_RESPONSE_BODY
+
+      it "returns true" do
+        expect(test_http_response).to receive(:nil?).and_return(false)
+        expect(test_http_response).to receive(:code).and_return(code)
+        expect(test_http_response).to receive(:body).and_return(body)
+
+        expect(subject.valid_response?(test_http_response)).to eq(true)
+      end
+    end
+
+    context "when the response is nil" do
+      it "reports the error" do
+        error = "Invalid response:"
+        expect(subject).to receive(:puts).with(error)
```

Comment:
This is what triggered me to realize the output was wonky in this case... the error message looks incomplete ending in a : with nothing after it.

Reply:
Updated